src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointProvider.cs

```diff
@@ -1,229 +1,107 @@
-        {
-            return Uri.TryCreate(endpoint, UriKind.Absolute, out var uriResult) &&
-                   (uriResult.Scheme == Uri.UriSchemeHttp || uriResult.Scheme == Uri.UriSchemeHttps);
+        private const string PreviewVersion = "1.0-preview";
+
+        private static readonly string ConnectionStringNotFound =
+            "No connection string was specified. " +
+            $"Please specify a configuration entry for {ServiceOptions.ConnectionStringDefaultKey}, " +
+            "or explicitly pass one using IServiceCollection.AddAzureSignalR(connectionString) in Startup.ConfigureServices.";
+
+        private readonly string _endpoint;
+        private readonly string _accessKey;
+        private readonly TimeSpan _accessTokenLifetime;
+        private readonly IServiceEndpointGenerator _generator;
+
+        public ServiceEndpointProvider(IOptions<ServiceOptions> options)
+        {
+            var connectionString = options.Value.ConnectionString;
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new ArgumentException(ConnectionStringNotFound);
+            }
+
+            _accessTokenLifetime = options.Value.AccessTokenLifetime;
+
+            string version;
+            int? port;
+            (_endpoint, _accessKey, version, port) = ConnectionStringParser.Parse(connectionString);
+
+            if (version == null || version == PreviewVersion)
+            {
+                _generator = new PreviewServiceEndpointGenerator(_endpoint, _accessKey);
+            }
+            else
+            {
+                _generator = new DefaultServiceEndpointGenerator(_endpoint, _accessKey, version, port);
+            }
+        }
+
+        public string GenerateClientAccessToken(string hubName, IEnumerable<Claim> claims = null, TimeSpan? lifetime = null)
+        {
+            if (string.IsNullOrEmpty(hubName))
+            {
+                throw new ArgumentNullException(nameof(hubName));
+            }
+
+            var audience = _generator.GetClientAudience(hubName);
+
+            return InternalGenerateAccessToken(audience, claims, lifetime ?? _accessTokenLifetime);
+        }
+
+        public string GenerateServerAccessToken<THub>(string userId, TimeSpan? lifetime = null) where THub : Hub
+        {
+            var audience = _generator.GetServerAudience(typeof(THub).Name);
+            var claims = userId != null ? new[] {new Claim(ClaimTypes.NameIdentifier, userId)} : null;
+
+            return InternalGenerateAccessToken(audience, claims, lifetime ?? _accessTokenLifetime);
+        }
+
+        public string GetClientEndpoint(string hubName, QueryString queryString)
+        {
+            if (string.IsNullOrEmpty(hubName))
+            {
+                throw new ArgumentNullException(nameof(hubName));
+            }
+
+            var endpoint = _generator.GetClientEndpoint(hubName);
+
+            if (queryString == QueryString.Empty)

```

Comment:
Also check null?

Reply:
`QueryString` is a struct (ValueType). It can never be `null`.